### PR TITLE
NAVAND-780: add initialLegIndex to RoutesUpdatedResult

### PR DESCRIPTION
**Pull request by dzinad**

### CODE REVIEW

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/MapboxRerouteController.kt [hunk 1]
```diff
@@ -123,6 +135,20 @@ internal class MapboxRerouteController @VisibleForTesting constructor(
...
+                    val legIndex = routeProgress.currentLegProgress?.legIndex
+                    val routeGeometryIndex = routeProgress.currentRouteGeometryIndex
+                    val legGeometryIndex = routeProgress.currentLegProgress?.geometryIndex
+                    ifNonNull(legIndex, legGeometryIndex) { legIndex, legGeometryIndex ->
+                        AlternativeDataProvider.getAlternativeLegIndex(
+                            RouteProgressData(legIndex, routeGeometryIndex, legGeometryIndex),
+                            alternativeMetadata
+                        )
+                    }
+                }
```
Comment (VysotskiVadim) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/MapboxRerouteController.kt:151]: @mapbox/navnative , how hard it is for you to pass us leg index of alternative route together with `getLocatedAlternativeRouteId`?

Comment (VysotskiVadim) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/MapboxRerouteController.kt:151]: > I don't think they track alternative route progress. 

Maybe it's time to start? 🙂 
I guess iOS have the same issue with multileg routes, @mapbox/navigation-ios , do you?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/RoutesObserver.kt [hunk 2]
```diff
@@ -30,10 +30,12 @@ fun interface RoutesObserver {
  * @param navigationRoutes list of currently maintained routes
  * @param ignoredRoutes list of alternative routes that were ignored
  *  because they are invalid for navigation. See [IgnoredRoute] for details.
+ * @param initialLegIndex the initial leg index of the primary route.
  * @param reason why the routes have been updated (re-route, refresh route, and etc.)
  */
 class RoutesUpdatedResult internal constructor(
     val navigationRoutes: List<NavigationRoute>,
+    val initialLegIndex: Int,
```
Comment (VysotskiVadim) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/RoutesObserver.kt:38]: how this number is supposed to be used? Should user always pass it to UI? What if route is refreshed on the second leg, will UI understand that it doesn't need to highlight initial leg as it's passed already?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/MapboxRerouteController.kt (hunk 1 above)
Comment (dzinad) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/MapboxRerouteController.kt:151]: iOS use native reroute controller and refresh controller, as far as I understand.
Started a conversation in Slack.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/RoutesObserver.kt (hunk 2 above)
Comment (VysotskiVadim) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/RoutesObserver.kt:38]: How is the SDK user supposed to calculate initial leg index for the case when user switches between alternatives clicking on them?

Comment (dzinad) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/RoutesObserver.kt:38]: Depending on how the new routes are set. They are probably set by the user. They do sth like:
```
val routes = listOf(newPrimaryRoute + mapboxNavigation.getNavigationRoutes().remove(newPrimaryRoute))
mapboxNavigation.setNavigationRoutes(routes)
```
Here we can only tell the user they should pass the leg index, we can't really do anything from our end.
All we see is new routes incoming with initialLegIndex = 0. Changing that even if we had enough data would be overriding user input, which just doesn't sound right.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/NavigationRerouteControllerV2.kt [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+import com.mapbox.navigation.core.MapboxNavigation
+import com.mapbox.navigation.core.trip.session.OffRouteObserver
+import com.mapbox.navigation.core.trip.session.RouteProgressObserver
+
+/**
+ * Reroute controller allows changing the reroute logic externally. Use [MapboxNavigation.rerouteController]
+ * to replace it.
+ */
+@UiThread
+interface NavigationRerouteControllerV2 : NavigationRerouteController {
```
Comment (dzinad) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/NavigationRerouteControllerV2.kt:17]: Will not be necessary after #7129.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/RoutesObserver.kt (hunk 2 above)
Comment (VysotskiVadim) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/RoutesObserver.kt:38]: @dzinad , as far as I remember from our conversation, you wanted to rename `initialLegIndex` to current leg index, didn't you?

Comment (VysotskiVadim) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/RoutesObserver.kt:38]: what do you think about exposing following API:

For the case when user manually switches to an alternative route by clicking on it: 
```
mapboxNavigation.switchToAlternative(alternativeId)
```
and for reroute case:
```
mapboxNavigation.deviateToAlternative(alternativeId)
```
?

**Comment by VysotskiVadim** [comment]
@dzinad , it feels like the case of switching from a multi-leg route to a single leg alternative and back is a Pandora's box for us, unfortunately we've found it when it's already open 😓 
The cases we need to address:
* reroute ✅ 
* manual switching to alternative, suggested [here](https://github.com/mapbox/mapbox-navigation-android/pull/7102#discussion_r1175222775)
* billing should respect leg index which was set, right now[ it seems that the SDK will start a new trip session when user switches back to multi-leg route](https://github.com/mapbox/mapbox-navigation-android/blob/99cd4ad116fc004e3742d10e9da48b6682fa06d4/libnavigation-core/src/main/java/com/mapbox/navigation/core/accounts/BillingController.kt#L284)

However I think that those cases are very rare, as we haven't received a single complain about other aspects of behaviour related to combination of routes of different legs count. 

Definitely not a blocker for this PR

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt [hunk 4]
```diff
@@ -511,6 +512,293 @@ class CoreRerouteTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::class.jav
...
+        ).getSuccessfulResultOrThrowException().routes
+        mapboxNavigation.setNavigationRoutes(routes)
+
+        mockLocationReplayerRule.loopUpdate(offRouteLocationUpdate, times = 5)
+
+        val rerouteResult = mapboxNavigation.routesUpdates().filter {
+            it.reason == RoutesExtra.ROUTES_UPDATE_REASON_REROUTE
+        }.first()
+        assertEquals(0, rerouteResult.initialLegIndex)
+    }
```
Comment (VysotskiVadim) [inline instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt:557]: What is the difference between this and `reroute_completes` test, except for the fact that `reroute_on_single_leg_route_without_alternatives` easier to read to me 🙂

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt [hunk 5]
```diff
@@ -511,6 +512,293 @@ class CoreRerouteTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::class.jav
...
+        mapboxNavigation.routeProgressUpdates().first()
+        mapboxNavigation.navigateNextRouteLeg()
+        mockLocationReplayerRule.loopUpdate(secondLegLocation, 100)
+        mapboxNavigation.routeProgressUpdates().filter { it.currentLegProgress?.legIndex == 1 }
+            .first()
+        mockLocationReplayerRule.loopUpdate(offRouteLocationUpdate, times = 5)
+
+        val rerouteResult = mapboxNavigation.routesUpdates().filter {
+            it.reason == RoutesExtra.ROUTES_UPDATE_REASON_REROUTE
+        }.first()
```
Comment (VysotskiVadim) [inline instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt:610]: How do you choose how many location updates need to be generated? As I see it's 100 for arrival and 5 for reroute.

What do you think about introducing a function, which updates location until some condition is met?

```suggestion
        stayOnPositionUntill(secondLegLocation) {
            mapboxNavigation.routeProgressUpdates()
                .filter { it.currentLegProgress?.legIndex == 1 }
                .first()
        }
       
       val rerouteResult = stayOnPositionUntill(offRouteLocationUpdate) {
               mapboxNavigation.routesUpdates().filter {
                   it.reason == RoutesExtra.ROUTES_UPDATE_REASON_REROUTE
                }.first()
       }
```

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/RoutesObserver.kt (hunk 2 above)
Comment (VysotskiVadim) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/RoutesObserver.kt:38]: But in this case we will need to also update `currentLegIndex` on route leg change

Comment (VysotskiVadim) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/RoutesObserver.kt:38]: Sharing more thoughts on this topic. Consider following case: user leaves current screen, but navigation keeps leaving, for example there is a running foreground service. When user enters the same screen, screens logic should be able to restore state based on navigation state. If user is on the second leg already, UI will highlight a wrong leg.

Comment (dzinad) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/RoutesObserver.kt:38]: > When user enters the same screen, screens logic should be able to restore state based on navigation state. If user is on the second leg already, UI will highlight a wrong leg.

Why?
Do you mean that the user will register `RoutesObserver` in `onStart` and potentially receive old data?

Comment (dzinad) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/directions/session/RoutesObserver.kt:38]: We can, but there's absolutely no guarantee that customers will switch to it.

**Comment by dzinad** [comment]
> billing should respect leg index which was set, right now[ it seems that the SDK will start a new trip session when user switches back to multi-leg route](https://github.com/mapbox/mapbox-navigation-android/blob/99cd4ad116fc004e3742d10e9da48b6682fa06d4/libnavigation-core/src/main/java/com/mapbox/navigation/core/accounts/BillingController.kt#L284)

Do you mean we should drop not only the origin, but `initialLegIndex + 1`?

File: instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt (hunk 5 above)
Comment (dzinad) [reply instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt:610]: The number of times doesn't really matter, because a new `loopUpdate` will stop the replayer and clear all the events. We just have to make multiple updates in order for NN to switch to this location.
But I like your suggestion.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/NavigationRerouteControllerV2.kt (hunk 3 above)
Comment (dzinad) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/NavigationRerouteControllerV2.kt:17]: Removed.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt [hunk 6]
```diff
@@ -1039,6 +1041,16 @@ class MapboxNavigation @VisibleForTesting internal constructor(
         CacheHandleWrapper.requestRoadGraphDataUpdate(navigator.cache, callback)
     }
 
+    /**
+     * Returns leg index the user is currently on. If no RouteProgress updates are available,
+     * returns the value passed as `initialLegIndex` parameter to `setNavigationRoutes`.
+     * If no routes are set, returns 0.
+     */
+    fun currentLegIndex(): Int {
```
Comment (VysotskiVadim) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:1049]: What do you think about nullable value?
```suggestion
     * If no routes are set, returns null.
     */
    fun currentLegIndex(): Int? {
```
plus make the route line accept null and fallback to 0, or turn off leg highlighting if no leg data is present?

Comment (dzinad) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:1049]: > plus make the route line accept null and fallback to 0, or turn off leg highlighting if no leg data is present?

Not so sure. Consider the case of route preview. Routes have not yet been set to MapboxNavigation, but they are set to `MapboxRouteLineApi`. So I think it's a valid case that no leg data is present when the method in invoked. And I don't think that we should turn off leg highlighting in this case.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/accounts/BillingController.kt [hunk 7]
```diff
@@ -275,13 +276,15 @@ internal class BillingController(
...
-    fun onExternalRouteSet(navigationRoute: NavigationRoute) {
+    fun onExternalRouteSet(navigationRoute: NavigationRoute, initialLegIndex: Int) {
         val runningSessionSkuId = getRunningOrPausedSessionSkuId()
         if (runningSessionSkuId == SessionSKUIdentifier.NAV2_SES_TRIP) {
             val currentRemainingWaypoints = getRemainingNonEVWaypointsOnRoute(
                 tripSession.getRouteProgress()
             )
-            val newWaypoints = getNonServerAddedWaypointsOnRoute(navigationRoute) { 1 }
+            val newWaypoints = getNonServerAddedWaypointsOnRoute(navigationRoute) {
+                countWaypointsUntilLegIndex(this, initialLegIndex) + 1
```
Comment (dzinad) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/accounts/BillingController.kt:286]: @LukasPaczos could you please tell me if you see any red flags with this particular change?
TLDR: we should drop not only the origin of a new route, but all the waypoints up to the one that corresponds to initialLegIndex (if it's 0, it's the origin).

Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/accounts/BillingController.kt:286]: I don't.
